main/lsp/requests/code_action.cc

```diff
@@ -29,8 +29,229 @@ vector<unique_ptr<TextDocumentEdit>> getEdits(const LSPConfiguration &config, co
+        if (!source.has_value()) {
+            return;
+        }
+        string newsrc;
+        if (auto sendResp = response->isSend()) {
+            edits[sendResp->receiverLoc] = newName;
+        }
+    }
+    void addSymbol(const core::SymbolRef symbol) override {
+        if (symbol.isMethod()) {
+            addSubclassRelatedMethods(gs, symbol.asMethodRef(), getQueue());
+        } else {
+            getQueue()->tryEnqueue(symbol);
+        }
+    }
+}; // CallSiteRenamer
+
+vector<unique_ptr<TextEdit>> moveMethod(const LSPConfiguration &config, const core::GlobalState &gs,
+                                        const core::lsp::DefinitionResponse *definition,
+                                        LSPTypecheckerInterface &typechecker, string_view newModuleName) {
+    auto moduleStart = fmt::format("module {}\n  extend T::Sig\n  ", newModuleName);
+    auto moduleEnd = "\nend";
+
+    auto fref = definition->termLoc.file();
+    auto &file = fref.data(gs);
+
+    auto trees = typechecker.getResolved({fref});
+    auto &rootTree = trees[0].tree;
+    auto beginLoc = core::Loc::offset2Pos(fref.data(gs), rootTree.loc().beginPos());
+
+    auto pos2Loc = [&](const auto &pos) {
+        auto loc = core::Loc::offset2Pos(fref.data(gs), pos);
+        return make_unique<Position>(loc.line - 1, loc.column - 1);
+    };
+
+    auto range2Loc = [&](const auto &pos) {
+        return make_unique<Range>(pos2Loc(pos.beginPos()), pos2Loc(pos.endPos()));
+    };
+
+    auto topOfTheFile = file.getLine(beginLoc.line);
+    auto sigAndMethodLocs = methodLocs(gs, rootTree, definition->symbol, fref);
+    if (!sigAndMethodLocs.has_value()) {
+        return {};
+    }
+    auto [sigLoc, methodLoc] = sigAndMethodLocs.value();
+    auto methodSource = copyMethodSource(gs, sigLoc, methodLoc, fref);
+
+    auto range = make_unique<Range>(make_unique<Position>(beginLoc.line - 1, 0),
+                                    make_unique<Position>(beginLoc.line - 1, topOfTheFile.length()));
+    auto replacement = fmt::format("{}{}{}\n\n{}", moduleStart, *methodSource, moduleEnd, topOfTheFile);
+    vector<unique_ptr<TextEdit>> res;
+    res.emplace_back(make_unique<TextEdit>(std::move(range), replacement));
+    res.emplace_back(make_unique<TextEdit>(range2Loc(sigLoc), ""));
+    res.emplace_back(make_unique<TextEdit>(range2Loc(methodLoc), ""));
+    return res;
+}
 } // namespace
 
+CodeActionTask::CodeActionTask(const LSPConfiguration &config, MessageId id, unique_ptr<CodeActionParams> params)
+    : LSPRequestTask(config, move(id), LSPMethod::TextDocumentCodeAction), params(move(params)) {}
+
+vector<unique_ptr<TextDocumentEdit>>
+CodeActionTask::getExtractMethodEdits(const LSPConfiguration &config, const core::GlobalState &gs,
+                                      const core::lsp::DefinitionResponse *definition,
+                                      LSPTypecheckerInterface &typechecker) {
+    ENFORCE(definition->symbol.isMethod());
```

Comment:
We should be able to delete this ENFORCE if we change `DefinitionResponse` to store a `MethodRef` as mentioned below.